src/Razor/src/Microsoft.CodeAnalysis.Razor.Workspaces/AutoInsert/IAutoInsertService.cs

```diff
@@ -0,0 +1,23 @@
+﻿// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the MIT license. See License.txt in the project root for license information.
+
+using System.Collections.Generic;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.CodeAnalysis.Razor.ProjectSystem;
+using Microsoft.VisualStudio.LanguageServer.Protocol;
+
+namespace Microsoft.CodeAnalysis.Razor.AutoInsert;
+
+internal interface IAutoInsertService
+{
+    IEnumerable<string> TriggerCharacters { get; }
+
+    ValueTask<InsertTextEdit?> TryResolveInsertionAsync(
+        IDocumentSnapshot documentSnapshot,
```

Comment:
From the code in this PR, it seems like `RazorCodeDocument` might be an easier fit for this parameter. Seems like the only thing that even happens with the `documentSnapshot` is getting the code document, or getting the source text. If you already have a code document, then source text is just the `.Source.Text` property, and these APIs seems like they could probably go back to being sync.

Same for `IOnAutoInsertProvider` below.

Reply:
Yep, looks like you are 100% correct here.